katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/backend/criteria/JpaCriteriaQueryBackend.java

```diff
@@ -0,0 +1,319 @@
+		}
+	}
+
+	@Override
+	public From<?, ?> getRoot() {
+		return root;
+	}
+
+	@Override
+	public void setOrder(List<Order> list) {
+		criteriaQuery.orderBy(list);
+	}
+
+	@Override
+	public List<Order> getOrderList() {
+		return new ArrayList<>(criteriaQuery.getOrderList());
+	}
+
+	@Override
+	public Order newSort(Expression<?> expr, Direction dir) {
+		if (dir == Direction.ASC) {
+			return cb.asc(expr);
+		} else {
+			return cb.desc(expr);
+		}
+	}
+
+	public CriteriaBuilder getCriteriaBuilder() {
+		return cb;
+	}
+
+	@Override
+	public void distinct() {
+		criteriaQuery.distinct(true);
+	}
+
+	public CriteriaQuery<T> getCriteriaQuery() {
+		return criteriaQuery;
+	}
+
+	@Override
+	public void addParentPredicate(MetaAttribute primaryKeyAttr) {
+		List<?> parentIds = queryImpl.getParentIds();
+		Path<Object> parentIdPath = parentFrom.get(primaryKeyAttr.getName());
+		addPredicate(parentIdPath.in(new ArrayList<Object>(parentIds)));
+	}
+
+	@Override
+	public boolean hasManyRootsFetchesOrJoins() {
+		return QueryUtil.hasManyRootsFetchesOrJoins(criteriaQuery);
+	}
+
+	@Override
+	public void addSelection(Expression<?> expression, String name) {
+		Selection<?> selection = criteriaQuery.getSelection();
+
+		List<Selection<?>> newSelection = new ArrayList<>();
+		if (selection != null) {
+			if (selection.isCompoundSelection()) {
+				newSelection.addAll(selection.getCompoundSelectionItems());
+			} else {
+				newSelection.add(selection);
+			}
+		}
+		newSelection.add(expression);
+		criteriaQuery.multiselect(newSelection);
+	}
+
+	@Override
+	public Expression<?> getExpression(Order order) {
+		return order.getExpression();
+	}
+
+	@Override
+	public boolean containsRelation(Expression<?> expression) {
+		return QueryUtil.containsRelation(expression);
+	}
+
+	public Predicate ilike(Expression<String> expr, String val) {
+		return cb.like(cb.lower(expr), val.toLowerCase());
+	}
+
+	private Predicate negateIfNeeded(Predicate p, FilterOperator fc) {
+		if (fc.equals(FilterOperator.NEQ))
+			return cb.not(p);
+		return p;
+	}
+
+	@Override
+	public Predicate buildPredicate(FilterOperator operator, MetaAttributePath attrPath, Object value) {
+		Expression<?> attr = getAttribute(attrPath);
+		return buildPredicate(operator, attr, value);
+	}
+
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	public Predicate buildPredicate(FilterOperator operator, Expression<?> expressionObj, Object value) {
+		Expression expression = expressionObj;
+
+		expression = handleConversions(expression, operator);
+
+		if (expression instanceof Predicate && expression.getJavaType() == Boolean.class
+				&& operator == FilterOperator.EQ) {
+			if (value.equals(Boolean.TRUE)) {
+				return (Predicate) expression;
+			} else {
+				return cb.not(expression);
+			}
+		}
+
+		return handle(expression, operator, value);
+
+	}
+
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	private Predicate handle(Expression expression, FilterOperator operator, Object value) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "handle" is 19 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)